File: plugins/codex-orchestrator/src/services/category-registry.ts
import { readFileSync } from "node:fs";
import type { CategoryDefinition, CategoryResolution } from "../types.ts";
import type { DelegationPreference } from "../types.ts";

type RawCategory = {
  intent?: string;
  preferred_role?: string;
  allowed_roles?: string[];
  write_policy?: string;
  requires_plan?: boolean;
  requires_spec_review?: boolean;
  requires_quality_review?: boolean;
  parallelism?: string;
  delegation_preference?: string;
  reuse_policy?: string;
  completion_contract?: string[];
};

type RawCategoryRecord = Record<string, RawCategory>;

function parseSimpleToml(source: string): RawCategoryRecord {
  const result: RawCategoryRecord = {};
  let currentSection: string | null = null;

  for (const rawLine of source.split(/\r?\n/u)) {
    const line = rawLine.trim();
    if (!line || line.startsWith("#")) continue;

    const sectionMatch = line.match(/^\[([^\]]+)\]$/u);
    if (sectionMatch) {
      currentSection = sectionMatch[1].trim();
      result[currentSection] = {};
      continue;
    }

    if (!currentSection) {
      throw new Error(`TOML key outside section: ${line}`);
    }

    const [rawKey, rawValue] = line.split("=", 2);
    if (!rawKey || rawValue === undefined) {
      throw new Error(`Invalid TOML line: ${line}`);
    }

    const key = rawKey.trim();
    const value = rawValue.trim();
    result[currentSection][key as keyof RawCategory] = parseTomlValue(value) as never;
  }

  return result;
}

function parseTomlValue(value: string): string | boolean | string[] {
  if (value === "true") return true;
  if (value === "false") return false;
  if (value.startsWith("[") && value.endsWith("]")) {
    const content = value.slice(1, -1).trim();
    if (!content) return [];
    return content
      .split(",")
      .map((entry) => entry.trim())
      .map((entry) => entry.replace(/^"(.*)"$/u, "$1"));
  }
  return value.replace(/^"(.*)"$/u, "$1");
}

export class CategoryRegistry {
  private readonly categories: Map<string, CategoryDefinition>;

  constructor(definitions: CategoryDefinition[]) {
    this.categories = new Map(definitions.map((definition) => [definition.id, definition]));
  }

  static fromToml(filePath: string): CategoryRegistry {
    const source = readFileSync(filePath, "utf8");
    const parsed = parseSimpleToml(source);
    const definitions = Object.entries(parsed).map(([id, raw]) => ({
      id,
      intent: raw.intent ?? id,
      preferredRole: raw.preferred_role ?? "default",
      allowedRoles: raw.allowed_roles ?? [raw.preferred_role ?? "default"],
      writePolicy: raw.write_policy ?? "read-only",
      requiresPlan: raw.requires_plan ?? false,
      requiresSpecReview: raw.requires_spec_review ?? false,
      requiresQualityReview: raw.requires_quality_review ?? false,
      parallelism: raw.parallelism ?? "single",
      delegationPreference: normalizeDelegationPreference(id, raw.delegation_preference),
      reusePolicy: raw.reuse_policy ?? "no_reuse",
      completionContract: raw.completion_contract ?? [],
    }));
    return new CategoryRegistry(definitions);
  }

  get(categoryId: string): CategoryDefinition | undefined {
    return this.categories.get(categoryId);
  }

  list(): CategoryDefinition[] {
    return Array.from(this.categories.values());
  }

  resolve(input: {
    title: string;
    description: string;
    explicitCategory?: string;
  }): CategoryResolution {
    if (input.explicitCategory) {
      const explicit = this.get(input.explicitCategory);
      if (!explicit) {
        throw new Error(`Unknown category: ${input.explicitCategory}`);
      }
      return {
        categoryId: explicit.id,
        reason: "explicit_category",
        category: explicit,
      };
    }

    const haystack = `${input.title}\n${input.description}`.toLowerCase();

    const planners = ["plan", "spec", "design", "architecture", "contract", "decompose"];
    if (planners.some((term) => haystack.includes(term))) {
      const category = this.require("plan");
      return { categoryId: category.id, reason: "planning_keywords", category };
    }

    const reviewers = ["review", "verify", "evaluation", "evaluator", "qa", "quality"];
    if (reviewers.some((term) => haystack.includes(term))) {
      const category = this.require("review");
      return { categoryId: category.id, reason: "review_keywords", category };
    }

    const researchers = ["research", "analyze", "analysis", "inspect", "scan", "find", "investigate", "map"];
    if (researchers.some((term) => haystack.includes(term))) {
      const category = this.require("research");
      return { categoryId: category.id, reason: "research_keywords", category };
    }

    const category = this.require("backend-impl");
    return { categoryId: category.id, reason: "default_backend_impl", category };
  }

  private require(categoryId: string): CategoryDefinition {
    const definition = this.get(categoryId);
    if (!definition) {
      throw new Error(`Missing required category definition: ${categoryId}`);
    }
    return definition;
  }
}

function normalizeDelegationPreference(categoryId: string, value: string | undefined): DelegationPreference {
  if (value === undefined) {
    throw new Error(`Category ${categoryId} is missing required delegation_preference`);
  }
  if (value === "parent-only" || value === "prefer-subagent" || value === "subagent-required") {
    return value;
  }
  throw new Error(`Category ${categoryId} has invalid delegation_preference: ${value}`);
}

File: plugins/codex-orchestrator/rust-cli/src/category_registry.rs
use std::{collections::HashMap, fs};

use anyhow::{anyhow, Context, Result};
use serde::Deserialize;

use crate::types::{CategoryDefinition, CategoryResolution, DelegationPreference};

#[derive(Debug, Deserialize)]
struct RawCategory {
    intent: Option<String>,
    preferred_role: Option<String>,
    allowed_roles: Option<Vec<String>>,
    write_policy: Option<String>,
    requires_plan: Option<bool>,
    requires_spec_review: Option<bool>,
    requires_quality_review: Option<bool>,
    parallelism: Option<String>,
    delegation_preference: Option<String>,
    reuse_policy: Option<String>,
    completion_contract: Option<Vec<String>>,
}

#[derive(Debug, Clone)]
pub struct CategoryRegistry {
    categories: HashMap<String, CategoryDefinition>,
}

impl CategoryRegistry {
    pub fn from_toml(file_path: &str) -> Result<Self> {
        let source = fs::read_to_string(file_path)
            .with_context(|| format!("failed to read categories TOML: {file_path}"))?;
        let parsed: HashMap<String, RawCategory> =
            toml::from_str(&source).context("failed to parse categories TOML")?;

        let definitions = parsed
            .into_iter()
            .map(|(id, raw)| {
                let preferred_role = raw.preferred_role.unwrap_or_else(|| "default".to_string());
                let allowed_roles = raw
                    .allowed_roles
                    .unwrap_or_else(|| vec![preferred_role.clone()]);
                Ok(CategoryDefinition {
                    id: id.clone(),
                    intent: raw.intent.unwrap_or_else(|| id.clone()),
                    preferred_role,
                    allowed_roles,
                    write_policy: raw.write_policy.unwrap_or_else(|| "read-only".to_string()),
                    requires_plan: raw.requires_plan.unwrap_or(false),
                    requires_spec_review: raw.requires_spec_review.unwrap_or(false),
                    requires_quality_review: raw.requires_quality_review.unwrap_or(false),
                    parallelism: raw.parallelism.unwrap_or_else(|| "single".to_string()),
                    delegation_preference: normalize_delegation_preference(
                        &id,
                        raw.delegation_preference.as_deref(),
                    )?,
                    reuse_policy: raw.reuse_policy.unwrap_or_else(|| "no_reuse".to_string()),
                    completion_contract: raw.completion_contract.unwrap_or_default(),
                })
            })
            .collect::<Result<Vec<_>>>()?;

        let categories = definitions
            .into_iter()
            .map(|definition| (definition.id.clone(), definition))
            .collect();

        Ok(Self { categories })
    }

    pub fn get(&self, category_id: &str) -> Option<&CategoryDefinition> {
        self.categories.get(category_id)
    }

    pub fn list(&self) -> Vec<CategoryDefinition> {
        self.categories.values().cloned().collect()
    }

    pub fn resolve(
        &self,
        title: &str,
        description: &str,
        explicit_category: Option<&str>,
    ) -> Result<CategoryResolution> {
        if let Some(explicit_category) = explicit_category {
            let explicit = self
                .get(explicit_category)
                .cloned()
                .ok_or_else(|| anyhow!("Unknown category: {explicit_category}"))?;
            return Ok(CategoryResolution {
                category_id: explicit.id.clone(),
                reason: "explicit_category".to_string(),
                category: explicit,
            });
        }

        let haystack = format!("{title}\n{description}").to_lowercase();

        let planners = ["plan", "spec", "design", "architecture", "contract", "decompose"];
        if planners.iter().any(|term| haystack.contains(term)) {
            let category = self.require("plan")?;
            return Ok(CategoryResolution {
                category_id: category.id.clone(),
                reason: "planning_keywords".to_string(),
                category,
            });
        }

        let reviewers = ["review", "verify", "evaluation", "evaluator", "qa", "quality"];
        if reviewers.iter().any(|term| haystack.contains(term)) {
            let category = self.require("review")?;
            return Ok(CategoryResolution {
                category_id: category.id.clone(),
                reason: "review_keywords".to_string(),
                category,
            });
        }

        let researchers = [
            "research",
            "analyze",
            "analysis",
            "inspect",
            "scan",
            "find",
            "investigate",
            "map",
        ];
        if researchers.iter().any(|term| haystack.contains(term)) {
            let category = self.require("research")?;
            return Ok(CategoryResolution {
                category_id: category.id.clone(),
                reason: "research_keywords".to_string(),
                category,
            });
        }

        let category = self.require("backend-impl")?;
        Ok(CategoryResolution {
            category_id: category.id.clone(),
            reason: "default_backend_impl".to_string(),
            category,
        })
    }

    fn require(&self, category_id: &str) -> Result<CategoryDefinition> {
        self.get(category_id)
            .cloned()
            .ok_or_else(|| anyhow!("Missing required category definition: {category_id}"))
    }
}

fn normalize_delegation_preference(
    category_id: &str,
    value: Option<&str>,
) -> Result<DelegationPreference> {
    match value {
        Some("parent-only") => Ok(DelegationPreference::ParentOnly),
        Some("prefer-subagent") => Ok(DelegationPreference::PreferSubagent),
        Some("subagent-required") => Ok(DelegationPreference::SubagentRequired),
        Some(other) => Err(anyhow!(
            "Category {category_id} has invalid delegation_preference: {other}"
        )),
        None => Err(anyhow!(
            "Category {category_id} is missing required delegation_preference"
        )),
    }
}
